### PR TITLE
Refactor user reference

### PR DESCRIPTION
**Pull request by matiascalegaris**
create a userreference struct that hold the index of the user array and another value to validate that we keep referencing the same instance instead of a reuse of the same index

### CODE REVIEW

File: Codigo/Modulo_UsUaRiOs.bas [hunk 1]
```diff
@@ -37,8 +37,47 @@ Attribute VB_Name = "UserMod"
 
 Option Explicit
 
+Public Function IsValidUserRef(ByRef UserRef As t_UserReference) As Boolean
+    IsValidUserRef = False
+    If UserRef.ArrayIndex <= 0 Then
```
Comment (morgolock) [inline Codigo/Modulo_UsUaRiOs.bas:42]: `or UserRef.ArrayIndex > UBOUND(Userlist)`

File: Codigo/Modulo_UsUaRiOs.bas [hunk 2]
```diff
@@ -37,8 +37,47 @@ Attribute VB_Name = "UserMod"
...
+    If UserList(UserRef.ArrayIndex).VersionId <> UserRef.VersionId Then
+        Exit Function
+    End If
+    IsValidUserRef = True
+End Function
+
+Public Function SetUserRef(ByRef UserRef As t_UserReference, ByVal index As Integer) As Boolean
+    SetUserRef = False
+    UserRef.ArrayIndex = Index
+    If Index <= 0 Then
```
Comment (morgolock) [inline Codigo/Modulo_UsUaRiOs.bas:54]: ` or index > UBOUND(Userlist)`

**Review by morgolock — CHANGES_REQUESTED** [review]

File: Codigo/MODULO_NPCs.bas [hunk 3]
```diff
@@ -473,17 +473,17 @@ Sub ResetNpcMainInfo(ByVal NpcIndex As Integer)
...
 150     NpcList(NpcIndex).Veneno = 0
 152     NpcList(NpcIndex).Desc = vbNullString
 154     NpcList(NpcIndex).NumDropQuest = 0
         
-156     If NpcList(NpcIndex).MaestroUser > 0 Then Call QuitarMascota(NpcList(NpcIndex).MaestroUser, NpcIndex)
+156     If IsValidUserRef(NpcList(npcIndex).MaestroUser) Then Call QuitarMascota(NpcList(npcIndex).MaestroUser.ArrayIndex, npcIndex)
 158     If NpcList(NpcIndex).MaestroNPC > 0 Then Call QuitarMascotaNpc(NpcList(NpcIndex).MaestroNPC)
 
-160     NpcList(NpcIndex).MaestroUser = 0
+160     Call SetUserRef(NpcList(npcIndex).MaestroUser, 0)
```
[the remark this answers is not in the document]
Comment (matiascalegaris) [reply Codigo/MODULO_NPCs.bas:486]: you clear this reference no other references and thats ok

File: Codigo/GameLogic.bas [hunk 4]
```diff
@@ -424,6 +434,8 @@ Public Sub ClearAttackerNpc(ByVal UserIndex As Integer)
 130     .flags.AtacadoPorNpc = 0
 132     .flags.NPCAtacado = 0
     End With
```
Comment (morgolock) [inline Codigo/GameLogic.bas:436]: missing `exit sub` before `ClearAttackerNpc_err`?

File: Codigo/ModEventos.bas [hunk 5]
```diff
@@ -264,33 +264,34 @@ Public Sub ForzarEvento(ByVal Tipo As Byte, ByVal Duracion As Byte, ByVal multi
         
         On Error GoTo ForzarEvento_Err
         
-
+        Dim tUser As t_UserReference
+        tUser = NameIndex(Quien)
 100     If Tipo > 3 Or Tipo < 1 Then
```
Comment (morgolock) [inline Codigo/ModEventos.bas:269]: missing `If IsValidUserRef(tUser) Then` ?

File: Codigo/Protocol_Writes.bas [hunk 6]
```diff
@@ -5654,7 +5654,7 @@ Public Sub WriteDebugLogResponse(ByVal UserIndex As Integer, ByVal debugType, By
         Next i
     ElseIf debugType = 1 Then
         'TODO- debug
-        Dim tIndex As Integer: tIndex = NameIndex(args(0))
+        Dim tIndex As Integer: tIndex = NameIndex(Args(0)).ArrayIndex
```
Comment (morgolock) [inline Codigo/Protocol_Writes.bas:5657]: `If IsValidUserRef(NameIndex(Args(0))) Then NameIndex(Args(0)).ArrayIndex`

**Review by morgolock — CHANGES_REQUESTED** [review]

File: Codigo/Protocol_Writes.bas (hunk 6 above)
Comment (matiascalegaris) [reply Codigo/Protocol_Writes.bas:5657]: no this is fine, we need to be able to debug a user ref even if its wrong... that might be the reason why we are debugging it in the first place

Comment (morgolock) [reply Codigo/Protocol_Writes.bas:5657]: ok

**Review by morgolock — APPROVED** [review]

**Pull request merged**